Fixed #35537 -- Use a namedtuple for email attachments and alternatives (#18261)
django/django: docs/topics/email.txt

  new connection is created when ``send()`` is called.

* ``attachments``: A list of attachments to put on the message. These can
  be either :class:`~email.mime.base.MIMEBase` instances, or a named tuple
  with attributes ``(filename, content, mimetype)``.

  .. versionchanged:: 5.2

>     ``attachments`` is now a list of named tuples as opposed to regular tuples.

Reviewer: ```suggestion
    In older versions, tuple items of ``attachments`` were regular tuples, 
    as opposed to named tuples.
```
Maybe 🤔 
We might also need to move this to the bottom of the `EmailMessage` definition

.. class:: EmailMultiAlternatives

>     A subclass of :class:`~django.core.mail.EmailMessage` that has an
>     additional ``attach_alternative()`` method and ``alternatives`` attribute
>     for including extra versions of the message body in the email. All the other
>     methods (including the class initialization) are inherited directly from
>     :class:`~django.core.mail.EmailMessage`.

Reviewer: ```suggestion
    A subclass of :class:`~django.core.mail.EmailMessage` that allows
    additional versions of the message body in the email via the
    ``attach_alternative()`` method. This directly inherits all methods
    (including the class initialization) from
    :class:`~django.core.mail.EmailMessage`.
```
This is based off [this discussion](https://github.com/django/django/pull/18278#discussion_r1644179202).

            msg.attach_alternative(html_content, "text/html")
            msg.send()

    .. attribute:: alternatives

        A list of named tuples with attributes ``(content, mimetype)``. This is
        particularly useful in tests::

            self.assertEqual(len(msg.alternatives), 1)
            self.assertEqual(msg.alternatives[0].content, html_content)

>         Alternatives should only be added using the ``attach_alternative()`` method.

Author: The change is backwards compatible, but I agree it'd be nice to flag seeing it's much nicer. I'll get it added.

            msg.attach_alternative(html_content, "text/html")
            msg.send()

    .. attribute:: alternatives

        A list of named tuples with attributes ``(content, mimetype)``. This is
        particularly useful in tests::

            self.assertEqual(len(msg.alternatives), 1)
            self.assertEqual(msg.alternatives[0].content, html_content)

        Alternatives should only be added using the ``attach_alternative()`` method.

        .. versionchanged:: 5.2

>             ``alternatives`` is now a list of named tuples as opposed to regular tuples.

Reviewer: ```suggestion
            In older versions, ``alternatives`` was a list of regular tuples, as opposed
            to named tuples.
```

            msg.attach_alternative(html_content, "text/html")
            msg.send()

    .. attribute:: alternatives

        A list of named tuples with attributes ``(content, mimetype)``. This is
        particularly useful in tests::

            self.assertEqual(len(msg.alternatives), 1)

>             self.assertEqual(msg.alternatives[0].content, html_content)

Reviewer: ```suggestion
            self.assertEqual(msg.alternatives[0].mime_type, "text/html")
```
Note that we've documented it to be `mimetype` rather than `mime_type`, I _think_ we prefer `mime_type` and should update the docs?

I also think adding tests doesn't hurt as at the very least it's easier to spot if we document the names wrong if we're using them in tests
```diff
--- a/tests/mail/tests.py
+++ b/tests/mail/tests.py
@@ -550,6 +550,29 @@ class MailTests(HeadersCheckMixin, SimpleTestCase):
         msg.attach("example.txt", "Text file content", "text/plain")
         self.assertIn(html_content, msg.message().as_string())

+    def test_attachments(self):
+        msg = EmailMessage()
+        file_name = "example.txt"
+        file_content = "Text file content"
+        mime_type = "text/plain"
+        msg.attach(file_name, file_content, mime_type)
+        self.assertEqual(msg.attachments[0][0], file_name)
+        self.assertEqual(msg.attachments[0].filename, file_name)
+        self.assertEqual(msg.attachments[0][1], file_content)
+        self.assertEqual(msg.attachments[0].content, file_content)
+        self.assertEqual(msg.attachments[0][1], mime_type)
+        self.assertEqual(msg.attachments[0].mime_type, mime_type)
+
+    def test_alternatives(self):
+        msg = EmailMultiAlternatives()
+        html_content = "<p>This is <strong>html</strong></p>"
+        mime_type = "text/html"
+        msg.attach_alternative(html_content, mime_type)
+        self.assertEqual(msg.alternatives[0][0], html_content)
+        self.assertEqual(msg.alternatives[0].content, html_content)
+        self.assertEqual(msg.alternatives[0][1], mime_type)
+        self.assertEqual(msg.alternatives[0].mime_type, mime_type)
+
     def test_none_body(self):
```

I also can't accept tickets I create, so feel free to accept this ticket: https://code.djangoproject.com/ticket/35537

Author: A quick search through the codebase suggests `mimetype` is more common than `mime_type`, although there isn't much consistency.

Definite +1 for adding tests though - I'll get those added!

Reviewer: Let's go for `mimetype` then 👍
